server/src/main/java/org/opensearch/index/remote/RemoteTranslogTransferTracker.java

```diff
@@ -0,0 +1,734 @@
+        }
+    }
+
+    public void recordDownloadStats(long bytesBefore, long downloadStartTime) {
+        long downloadEndTime = System.nanoTime();
+        long downloadEndTimeMs = System.currentTimeMillis();
+        long durationInMillis = (downloadEndTime - downloadStartTime) / 1_000_000L;
+        long bytesDownloaded = getDownloadBytesSucceeded() - bytesBefore;
+
+        setLastSuccessfulDownloadTimestamp(downloadEndTimeMs);
+
+        // We update the duration at the end of successfully downloading all of metadata, .tlog, .ckp
+        // files because this is not a file-level metric but a sync-level metric.
+        // This also ensures the bytes per sec moving average can be correlated.
+        addDownloadTimeInMillis(durationInMillis);
+        updateDownloadBytesMovingAverage(bytesDownloaded);
+        updateDownloadTimeMovingAverage(durationInMillis);
+        if (durationInMillis > 0) {
+            updateDownloadBytesPerSecMovingAverage(bytesDownloaded * 1_000L / durationInMillis);
+        }
+    }
+
+    /**
+     * Gets the tracker's state as seen in the stats API
+     * @return Stats object with the tracker's stats
+     */
+    public RemoteTranslogTransferTracker.Stats stats() {
+        return new RemoteTranslogTransferTracker.Stats(
+            shardId,
+            lastSuccessfulUploadTimestamp.get(),
+            totalUploadsStarted.get(),
+            totalUploadsSucceeded.get(),
+            totalUploadsFailed.get(),
+            uploadBytesStarted.get(),
+            uploadBytesSucceeded.get(),
+            uploadBytesFailed.get(),
+            totalUploadTimeInMillis.get(),
+            uploadBytesMovingAverageReference.get().getAverage(),
+            uploadBytesPerSecMovingAverageReference.get().getAverage(),
+            uploadTimeMsMovingAverageReference.get().getAverage(),
+            lastSuccessfulDownloadTimestamp.get(),
+            totalDownloadsSucceeded.get(),
+            downloadBytesSucceeded.get(),
+            totalDownloadTimeInMillis.get(),
+            downloadBytesMovingAverageReference.get().getAverage(),
+            downloadBytesPerSecMovingAverageReference.get().getAverage(),
+            downloadTimeMsMovingAverageReference.get().getAverage()
+        );
+    }
+
+    /**
+     * Validates that the sum of successful operations, failed operations, and the number of operations to add (irrespective of failed/successful) does not exceed the number of operations originally started
+     * @param startedCount Number of operations started
+     * @param failedCount Number of operations failed
+     * @param succeededCount Number of operations successful
+     * @param countToAdd Number of operations to add
+     */
+    private void checkTotal(long startedCount, long failedCount, long succeededCount, long countToAdd) {
+        long delta = startedCount - (failedCount + succeededCount + countToAdd);
+        assert delta >= 0 : "Sum of failure count ("
```

Comment:
Looks like delta should be equal to 0 always. Are there any cases when this can be greater than 0?

Reply:
Based on the way this method is currently invoked, there wouldn't be a case when this will be > 0. 

But, a valid scenario is if for example 10 uploads started and then we were notified that 2 failed and 4 succeeded but are still awaiting status of 4 pending uploads. In this case, the `delta` being > 0 is a valid case.

Reply:
Got it. This can happen during concurrent uploads.

Reply:
@ashking94 changing this to `== 0` is failing multiple tests in `RemoteSegmentTransferTrackerTests`. Can we track this requirement of changing this constraint to `== 0` in a separate issue to avoid lumping any further changes with the translog stats changes? The tests in `RemoteSegmentTransferTrackerTests` need to be refactored to depict the runtime state.